### PR TITLE
Refactoring: entities-table / entity-subrecord

### PR DESCRIPTION
**Pull request by sleidig**
Open TODOs
- [x] fix remaining tests
- [x] sort overwrite (e.g. TodoListComponent) doesn't work
- [x] trying to inline-edit a note (e.g. in child details) always opens the popup
- [x] when clicking on a from field saving after inline editing does not reset the unsaved changes state (UnsavedChangesService) so the app still thinks there are unsaved changes
- [x] Editing and saving in a popup (note details or row details) does not update the value in the table after closed --> (I couldn't reproduce this, @TheSlimvReal )
- [x] Readonly function is not shown in popup (e.g. health check BMI)
- [x] More properties are shown in popup (e.g. health check also shows child id form) --> reimplement hideFromTable / hideFromForm

potential further refactoring
- [ ] DataSource extension to handle all filtering? (replacing FilterService?) --> see https://blog.angular-university.io/angular-material-data-table/  --> #2161

### CODE REVIEW

File: src/app/core/common-components/entities-table/list-paginator/list-paginator.component.scss [hunk 1]
```diff
@@ -1,4 +1,4 @@
-@use "variables/sizes";
+@use "../../../../../styles/variables/sizes";
```
Comment (sleidig) [inline src/app/core/common-components/entities-table/list-paginator/list-paginator.component.scss:1]:
```suggestion
@use "variables/sizes";
```

File: src/app/core/common-components/entities-table/entity-inline-edit-actions/entity-inline-edit-actions.component.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import { FaIconComponent } from "@fortawesome/angular-fontawesome";
+import { MatButtonModule } from "@angular/material/button";
+import { TableRow } from "../entities-table.component";
+import { Entity } from "../../../entity/model/entity";
+import { InvalidFormFieldError } from "../../entity-form/invalid-form-field.error";
+import { EntityFormService } from "../../entity-form/entity-form.service";
+import { AlertService } from "../../../alerts/alert.service";
+import { EntityActionsService } from "../../../entity/entity-actions/entity-actions.service";
+
+@Component({
```
Comment (sleidig) [inline src/app/core/common-components/entities-table/entity-inline-edit-actions/entity-inline-edit-actions.component.ts:14]:
```suggestion
/**
 * Buttons to edit an (entities-table) row inline, handling the necessary logic and UI buttons.
 */
@Component({
```

File: src/app/core/common-components/entities-table/entities-table.component.ts [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    if (!row.formGroup || row.formGroup.disabled) {
+      this.showEntity(row.record);
+    }
+    this.rowClick.emit(row.record);
+  }
+
+  showEntity(entity: T) {
+    switch (this.clickMode) {
+      case "popup":
+        this.formDialog.openFormPopup(entity, this.columnsToDisplay); // TODO this.formDialog.openFormPopup(entity, this._columns)
```
Comment (sleidig) [inline src/app/core/common-components/entities-table/entities-table.component.ts:273]: we do not correctly support the following flags yet:
- hideFromTable?: boolean;
- hideFromForm?: boolean;
- forTable?: boolean;

Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:273]: --> discuss: where + how do we need these concepts; let's try to simplify and clearly specify this before re-adding the logic.

(e.g. todos-related-to-entity use the process)

File: src/app/core/entity-details/related-entities/related-entities.component.ts [hunk 4]
```diff
@@ -36,54 +45,75 @@ export class RelatedEntitiesComponent<E extends Entity> implements OnInit {
...
     if (this.showInactive === undefined) {
       this.showInactive = this.entity.anonymized;
     }
+  }
 
-    this.isLoading = false;
+  private updateSubscription: Subscription;
+
+  protected listenToEntityUpdates() {
+    if (!this.updateSubscription && this.entityCtr) {
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/related-entities/related-entities.component.ts:109]: Do we actually need this check if the function is only once called in `ngOnInit`?

File: src/app/core/entity-details/related-entities/related-entities.component.ts [hunk 5]
```diff
@@ -96,4 +126,27 @@ export class RelatedEntitiesComponent<E extends Entity> implements OnInit {
...
+          return false;
+        }
+
+        const numericValue = ScreenSize[column?.visibleFrom];
+        if (numericValue === undefined) {
+          return true;
+        }
+        return this.screenWidthObserver.currentScreenSize() >= numericValue;
+      })
+      .map((c) => toFormFieldConfig(c).id);
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/related-entities/related-entities.component.ts:150]: Does `toFormFieldConfig` really need to be called 3 times in this function? I think running this once in the beginning would make the code much simpler.

File: src/app/core/entity-details/related-time-period-entities/related-time-period-entities.component.ts [hunk 6]
```diff
@@ -64,11 +66,12 @@ export class RelatedTimePeriodEntitiesComponent<E extends TimePeriod>
   hasCurrentlyActiveEntry: boolean;
 
   async ngOnInit() {
-    this.onIsActiveFilterChange(this.showInactive);
     await super.initData();
+    super.listenToEntityUpdates();
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/related-time-period-entities/related-time-period-entities.component.ts:70]:
```suggestion
    await super.ngOnInit();
```

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
Really good refactoring. Most comments are quite minor. 

On functionality testing I discovered some bugs:
- [x] trying to inline-edit a note (e.g. in child details) always opens the popup when clicking on a from field
- [ ] saving after inline editing does not reset the unsaved changes state (`UnsavedChangesService`) so the app still thinks there are unsaved changes
- [ ] Editing and saving in a popup (note details or row details) does not update the value in the table after closed
- [ ] Readonly function is not shown in popup (e.g. health check BMI)
- [ ] More properties are shown in popup (e.g. health check also shows child id form)

File: src/app/child-dev-project/schools/child-school-overview/child-school-overview.component.ts [hunk 7]
```diff
@@ -10,14 +10,19 @@ import { MatSlideToggleModule } from "@angular/material/slide-toggle";
...
+import { EntityMapperService } from "../../../core/entity/entity-mapper/entity-mapper.service";
+import { EntityRegistry } from "../../../core/entity/database-entity.decorator";
+import { ScreenWidthObserver } from "../../../utils/media/screen-size-observer.service";
+import { UntilDestroy } from "@ngneat/until-destroy";
 
 // TODO: once schema-generated indices are available (#262), remove this component and use its generic super class directly
 @DynamicComponent("ChildSchoolOverview")
 @DynamicComponent("PreviousSchools")
 @DynamicComponent("ChildrenOverview")
+@UntilDestroy()
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/schools/child-school-overview/child-school-overview.component.ts:25]: I tested it again with proper analysis and indeed, the `@UntilDestroy` is not needed here but taken from the base class. Thanks for the suggestion

File: src/app/core/common-components/entities-table/entities-table.component.html [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    <!--
+      BULK SELECT
+    -->
+    <ng-container [matColumnDef]="ACTIONCOLUMN_SELECT">
+      <th mat-header-cell *matHeaderCellDef style="width: 0"></th>
+
+      <td mat-cell *matCellDef="let row">
+        <mat-checkbox
+          *ngIf="_selectable"
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.html:50]: no good reason. In fact the ngIf is not actually needed for display control (that's through columnsToDisplay anyway), so I am removing it completely

File: src/app/core/common-components/entities-table/entities-table.component.ts [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    );
+    const entityColumns = this.entityType?.schema
+      ? [...this.entityType.schema.entries()].map(
+          ([id, field]) => ({ ...field, id }) as FormFieldConfig,
+        )
+      : [];
+
+    const allColumns = [...entityColumns, ...this._customColumns];
+    this._columns = allColumns
+      // remove duplicates (keep customColumn = last in array)
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:107]: added better comment:
```
      //   if there is a customColumn for a field from entity config, the custom FormFieldConfig takes precedent (which is the latter item with the same id in array)
```

File: src/app/core/entity-details/related-entities/related-entities.component.ts (hunk 4 above)
Comment (sleidig) [reply src/app/core/entity-details/related-entities/related-entities.component.ts:109]: no, in most of these we don't. Thanks!

File: src/app/core/entity-details/related-entities/related-entities.component.ts (hunk 5 above)
Comment (sleidig) [reply src/app/core/entity-details/related-entities/related-entities.component.ts:150]: I have no clue how this completely pointless .find ended up here. Simplified now :sweat_smile:

File: src/app/child-dev-project/notes/notes-manager/notes-manager.component.ts [hunk 10]
```diff
@@ -56,8 +56,7 @@ export class NotesManagerComponent implements OnInit {
 
   config: EntityListConfig;
   noteConstructor = Note;
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/notes/notes-manager/notes-manager.component.ts:58]: the NotesManagerComponen.subscribeEntityUpdates is actually adding this here and seems to work for the combination. (I did also test and ensure that it doesn't colide with the updates that EntityList already does)

File: src/app/core/common-components/entities-table/entities-table.component.ts (hunk 3 above)
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:273]: we discussed to move towards a different config structure in the future:
- config for list columns
- config for details fields (popup or fullscreen) separately
- config for entity, including custom (run-time calculated) field configs

image


Might only refactor things towards this at a later point, however.

File: src/app/core/common-components/entities-table/entities-table.component.ts [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    if (
+      sortByColumn?.viewComponent === "DisplayDate" ||
+      sortByColumn?.viewComponent === "DisplayMonth" ||
+      [
+        DateDatatype.dataType,
+        DateOnlyDatatype.dataType,
+        DateWithAgeDatatype.dataType,
+        MonthDatatype.dataType,
+      ].includes(sortByColumn?.dataType)
+    ) {
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:308]: nice idea to use inheritance. This actually works :+1:

File: src/app/core/common-components/entities-table/entities-table.component.ts
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:1]: I do agree. I think the idea to put filter (and sorting) logic into a custom DataSource implementation would help separate things further. But I'd suggest to postpone this for a few weeks to make progress on user-facing features

**Comment by tomwwinter** [comment]
I took a look into the new entities-table.component, it's still a very big component but it looks good so far. Let's try!

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
Thanks for the cleanups. Code looks really good.

I discovered a few more bugs though:

- [x] Default sort ist not by first column in children list
- [x] Todos are sorted ascending by deadline date
- [x] clicking on a Todo opens 404 page and popup (should only open popup)
- [x] Deleting a note in notes manager and pressing "undo" only shows the recovered note after re-entering the page

**Comment by sleidig** [comment]
> Todos are sorted ascending by deadline date

This is actually intentionally configured like this (assuming that users should see those tasks that are the longest overdue / closest to due that first)

File: src/app/core/common-components/entities-table/entities-table.component.ts [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+    if (value === this._showInactive) {
+      return;
+    }
+
+    this._showInactive = value;
+    this.updateFilteredData();
+    this.showInactiveChange.emit(value);
+  }
+  _showInactive: boolean = false;
+  @Output() showInactiveChange = new EventEmitter<boolean>();
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/common-components/entities-table/entities-table.component.ts:331]: good that you are pointing this out. We should use this to switch the "highlight current child-school-relations" logic (which is only active if inactive records are displayed). That is currently still a UI regression ...

File: src/app/core/config/config-fix.ts [hunk 13]
```diff
@@ -168,12 +168,6 @@ export const defaultJsonConfig = {
       "title": $localize`:Title for notes overview:Notes & Reports`,
       "includeEventNotes": false,
       "showEventNotesToggle": true,
-      "columns": [
-        {
-          "id": "children",
-          "noSorting": true
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/config/config-fix.ts:174]: yes, I think this should be handled by entities-table. We probably haven't disabled it for every relevant case currently.

It seems like sorting on Note.authors works (which is also an entity-array)? Any idea around that?

[the remark this answers is not in the document]
Comment (TheSlimvReal) [reply src/app/core/config/config-fix.ts:174]: Yes, authors are users and their ID is mostly the same as the display name.

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
Seems to all work well now. There are just two more minor things I discovered.

- [ ] Inline editing "cancel" does not correctly reset `unsavedChanges` state
- [ ] The Task list does by default not show the archived tasks anymore. This results in the weird behavior that when activating the "completed" filter no tasks are shown until the "include archived" is also activated. This is probably something that most users will not figure out.

File: src/app/core/config/config-fix.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/config/config-fix.ts:174]: :+1: will deactivate for all entity ref dataTypes then

**Review by TheSlimvReal — APPROVED** [review]
Great, thanks for also implementing the sorting automation.

**Pull request merged**